### ISSUE TITLE
why use BOS and EOS token and not only the EOS?

### ISSUE THREAD

**Issue by CohleM** [jingyaogong/minimind#130]
In the `model/dataset.py`

line 29

` text = f"{self.tokenizer.bos_token}{str(sample['text'])}{self.tokenizer.eos_token}"` 

you prepend bos token and then append the eos token to the end of text and pad the tokens if shorter than the context length. Is there any reason for doing that? 

and why don't you just append the EOS token to the end of the text and fill it with the next sequence rather than padding it?

**Comment by jingyaogong** [comment]
1. In pre-training, there are practices that use both [BOS] and [EOS] tokens to define text boundaries, as well as approaches that omit [BOS] and only use text + [EOS]. We chose the latter, as there is no strict rule. On the other hand, in the instruction fine-tuning template for chat, dialogue roles start with [BOS] + user/assistant/system. Therefore, we maintained the consistency of starting with [BOS] during pre-training.

2. The padding after [EOS] is masked to avoid participating in loss calculation, preventing the model from learning invalid information. If unrelated segments are dynamically concatenated, it introduces complexity and may create false contextual associations.

3. Independent sample processing (non-concatenation) aligns better with real-world inference scenarios (e.g., single-document generation), avoiding cross-text semantic contamination.

**Comment by CohleM** [comment]
Thank you for the response. My concern was with the addition of pad tokens, as it simply wastes the extra computation which could have been used if the sequences were simply concatenated. 


For

> 3. Independent sample processing (non-concatenation) aligns better with real-world inference scenarios (e.g., single-document generation), avoiding cross-text semantic contamination.

do you have any data that backs this claim? I'm currently exploring which process (concatenation or non-concatenation) better suites pre-training and would really appreciate your help.

**Comment by jingyaogong** [comment]
First, the consensus is that the computational waste caused by padding is unavoidable, with only the degree of severity varying.  
If it is [Document A content...<EOS>][Document B content...<EOS>][Document C content...<EOS>] ➜ processed as a single sample during training.

I believe that if the input sequence contains unrelated texts (e.g., news + code), the model will simultaneously learn false associations between them (e.g., "stock rise" followed by `print("Hello")`), leading to confusion in distribution modeling. However, since EOS serves as a delimiter, it allows the model to infer explicit boundaries.

GPT-3's *"Language Models are Few-Shot Learners"* mentions that using a reasonable context window and clear separators indeed helps the model better understand document boundaries and reduce perplexity. *"Scaling Laws for Neural Language Models"* points out that noise or irrelevant information within long contexts may increase perplexity, so it is more beneficial to avoid concatenating documents during training and keep each sample independent.

Although EOS exists, it still inevitably slightly increases the model's perplexity (only a matter of degree compared to independent data). However, I have yet to find objective data supporting this view, as all papers only provide subjective evaluations.

Nevertheless, compared to the increase in training efficiency, the slight cost is, in my opinion, entirely acceptable.
